### DISCUSSION TITLE
7.0.11-5 produces apparently invalid DDS images

### DISCUSSION THREAD

**Discussion by wmassingham** [ImageMagick/ImageMagick#3486] [Help] [answered]
I'm trying to convert png images to dds, however after conversion, they fail to open in GIMP or DDS Viewer. GIMP complains about unexpected EOF and DDS viewer crashes. Am I doing something wrong, or is DDS conversion broken in this version? Is there a version I should roll back to?

Example image: https://homepages.cae.wisc.edu/~ece533/images/fruits.png

Command: `magick.exe convert fruits.png fruits.dds`

I've also tried specifying `-format dds`, `-define dds:compression=dxt1` (and dxt5) with no change. 

`magick.exe identify` says `fruits.dds DDS 512x512 512x512+0+0 8-bit sRGB 131072B 0.000u 0:00.000` so it *appears* valid. 

Further, trying to convert it back to e.g. bmp with `magick.exe convert fruits.dds fruits.bmp` says "convert: no images defined `fruits.bmp' @ error/convert.c/ConvertImageCommand/3322."

I'm running on Windows 10 version 10.0.19042.870, 64-bit.
imagemagick version info, as installed from ImageMagick-7.0.11-5-Q16-HDRI-x64-dll.exe:

```
Version: ImageMagick 7.0.11-5 Q16 x64 2021-03-28 https://imagemagick.org
Copyright: (C) 1999-2021 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Visual C++: 192829912
Features: Cipher DPC HDRI Modules OpenCL OpenMP(2.0)
Delegates (built-in): bzlib cairo flif freetype gslib heic jng jp2 jpeg jxl lcms lqr lzma openexr pangocairo png ps raw rsvg tiff webp xml zip zlib
```

**Comment by wmassingham** [comment]
Well, I have WSL enabled so I tried version 6.9.10-23 from the Ubuntu repos, and in that version just running `convert fruits.png fruits.dds` worked as expected. So I guess something is broken in 7.0.11-5.

**Comment by fmw42** [comment]
In IM 7, use magick, not magick  convert. Also perhaps

`magick.exe fruits.png DDS:fruits.dds`

Sorry, I am not a Microsoft user. So do not know much specific about DDS.

**Comment by wmassingham** [comment]
I just tried that too, no difference.

But I also just installed ImageMagick-6.9.12-5-Q16-HDRI-x64-dll.exe on Windows and with that version, the same `convert fruits.png fruits.dds` that works under Ubuntu _fails_ on Windows, so it's not the change from 6 to 7 that broke after all, it's either something between 6.9.10-23 and 6.9.12-5, or something due to using the Windows build.

**Comment by bulbaka** [comment]
@wmassingham,
have tested your fruits.png to .dds conversion.

Environment:
ImageMagick 7.0.11-4 Q16 x64 2021-03-20 from ImageMagick-7.0.11-4-portable-Q16-x64.zip @
Win10 x64 20H2 Build 19042.867.

code for Windows .cmd:
````
magick fruits.png -define dds:mipmaps=0 -define dds:compression=dxt1 fruits_dxt1_mm0.dds
magick fruits.png -define dds:mipmaps=4 -define dds:compression=dxt1 fruits_dxt1_mm4.dds
````

Everything seems good and without any issues while viewing (Imagine viewer) and opening (Paint.NET) noticed so far.

Try to get extended information ("-verbose") about involved .png and .dds images (code for Windows .cmd):
````
for %%d in (*.dds) do (
identify -verbose %%d > %%d_info.txt
)
for %%p in (*.png) do (
identify -verbose %%p > %%p_info.txt
)
````

Edit 1:
To get dxt5 compression try this ("PNG32:" adds "Type: TrueColorAlpha" to the source .png to bring formal sense for dxt5)
(code for Windows .cmd):
````
magick fruits.png PNG32:fruits_32.png
magick fruits_32.png -define dds:mipmaps=0 -define dds:compression=dxt5 fruits_dxt5_mm0.dds
magick fruits_32.png -define dds:mipmaps=4 -define dds:compression=dxt5 fruits_dxt5_mm4.dds
````

PS Please find my converting results attached for testing
[fruits.png.zip](https://github.com/ImageMagick/ImageMagick/files/6242002/fruits.png.zip)
[fruits_32.png.zip](https://github.com/ImageMagick/ImageMagick/files/6242003/fruits_32.png.zip)
[fruits_dxt1_mm0.dds.zip](https://github.com/ImageMagick/ImageMagick/files/6242004/fruits_dxt1_mm0.dds.zip)
[fruits_dxt1_mm4.dds.zip](https://github.com/ImageMagick/ImageMagick/files/6242005/fruits_dxt1_mm4.dds.zip)
[fruits_dxt5_mm0.dds.zip](https://github.com/ImageMagick/ImageMagick/files/6242006/fruits_dxt5_mm0.dds.zip)
[fruits_dxt5_mm4.dds.zip](https://github.com/ImageMagick/ImageMagick/files/6242007/fruits_dxt5_mm4.dds.zip)
[Edit 2: have downloaded and installed DDS viewer, mentioned by OP, and it handles my .dds pretty well, without crashes.
By the way, what an awful example of software for today. Immediately uninstalled it after a single launch.
Thanks to the authors but this tool days have gone.
Please, try something else for the .dds viewing, for instance, [Imagine](http://www.nyam.pe.kr/dev/imagine/)]

**Comment by wmassingham** [comment, marked as answer]
Interesting. Setting `mipmaps=0` was the only one that produced a valid result for me. Setting `mipmaps=4` or omitting the option produced an image I can't open. I've attached my output.

Thanks for the pointer to Imagine, I was just using DDS Viewer as a second opinion to GIMP. Imagine *is* able to open all files, and the full-res image is correct, but the mipmaps are garbage. This explains the behavior I was seeing.

[fruits.zip](https://github.com/ImageMagick/ImageMagick/files/6245010/fruits.zip)

**Reply by bulbaka** [reply]
> In computer graphics, mipmaps (also MIP maps) or pyramids are pre-calculated, optimized sequences of images, each of which is a progressively lower resolution representation of the previous. The height and width of each image, or level, in the mipmap is a power of two smaller than the previous level... They are intended to increase rendering speed and reduce aliasing artifacts... [wikipedia](https://en.wikipedia.org/wiki/Mipmap)

As long as without apparent setting, ImageMagick by default produces .dds with mipmaps (to some extent just a multi-page image), it's been my first suggestion, that your software can't handle those, and the `mipmaps=0` (which tells ImageMagick to make a single-image .dds) option seemed the first idea worth trying.

Glad if this helps.

Edit 1.
Took a look into your fruits.zip
Surprisingly, when I try to open your .dds with mimaps<>0 my Paint.NET can't handle them as well, while Imagine allows to see them with only the first level texture visible, rendering all the downscaled textures as actual garbage.
Something is wrong there.

Edit 2.
I've downloaded ImageMagick-7.0.11-5-portable-Q16-x64.zip to run the above scripts:
````@echo off

:: dxt1 compression
magick fruits.png -define dds:mipmaps=0 -define dds:compression=dxt1 fruits_dxt1_mm0.dds
magick fruits.png -define dds:mipmaps=4 -define dds:compression=dxt1 fruits_dxt1_mm4.dds

:: dxt5 compression
magick fruits.png PNG32:fruits_32.png
magick fruits_32.png -define dds:mipmaps=0 -define dds:compression=dxt5 fruits_dxt5_mm0.dds
magick fruits_32.png -define dds:mipmaps=4 -define dds:compression=dxt5 fruits_dxt5_mm4.dds
````

With this version, I'm getting practically the same results as yours.

So now I can **confirm the issue:
ImageMagick-7.0.11-5(-portable-Q16-x64) is buggy, it produces .dds with corrupt mipmaps unless '-define dds:mipmaps=0' option is set.**

In its turn `identify -verbose` returns zero information processing those corrupt .dds.

There's nothing like that with the previous ImageMagick-7.0.11-4(-portable-Q16-x64).

Edit 3.
Apparently, this thread belongs to the ImageMagick Issues section: 
https://github.com/ImageMagick/ImageMagick/issues

**Reply by wmassingham** [reply]
Thanks. I've opened issue https://github.com/ImageMagick/ImageMagick/issues/3492.
